### DISCUSSION TITLE
Automatic deployment does not happen after adding vercel.json

### DISCUSSION THREAD

**Discussion by muezz** [vercel/community#4424] [Help] [answered]
### Summary

I have hosted a Next13 app with Vercel such that any commits to `dev` are deployed to `preview` and changes to `main` are deployed to `production`. This was done automatically but now I want to set up a cron job (see below). I added the `vercel.json` file as per the docs but the automatic deployment has stopped.

I am sure the inclusion of this file is overriding some default settings that I need to specify again in this file. What are those settings?

### Steps to Reproduce
```json
{
    "crons": [
        {
            "path": "/api/v1/email/reports/hosting",
            "schedule": "0 */5 * ? * *"
        }
    ]
}
```

**Comment by amyegan** [comment]
Hey @muezz! ~~Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that we can use to help with debugging?~~

Scratch that, I see the problem! If you [validate the cron expression](https://vercel.com/docs/cron-jobs#validate-cron-expressions), you'll see that `?` is not a valid character. The error seems to be preventing automatic builds. If you change the expression to `0 */5 * * *`, then it will start working again 🙏

**Reply by muezz** [reply]
@amyegan I tried this cron expression and it still did not work.

TLDR: If I remove this `vercel.json` file, the automatic deployment works just fine. It stops working as soon as I add it.

**Reply by amyegan** [reply]
You should always validate the cron expression at https://vercel.com/docs/cron-jobs#validate-cron-expressions or with a tool like [crontab guru](https://crontab.guru/). If there is a problem with the expression, the build will fail.

Assuming you're on a Pro account, it should work with this in `vercel.json`:
```
{
  "crons": [
    {
      "path": "/api/v1/email/reports/hosting",
      "schedule": "0 */5 * * *"
    }
  ]
}
```

**Reply by muezz** [reply]
@amyegan I did validate it once you pointed out the error/mistake. Despite that, my next app is not deploying. My suspicion is that there needs to be something else in this `vercel.json` file.

**Reply by amyegan** [reply]
Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to help with debugging? I wasn't able to reproduce this problem with my test app.

**Reply by muezz** [reply]
Not sure what to put in the minimum reproducible example.

The app is a pretty generic Next13 app. It is being deployed automatically whenever there is a commit to the `dev` or `main` branch. All that was done automatically.

The automatic deployment is **not** even triggered now and the only difference is that I have added a file called `vercel.json` in the root of the project directory. The only content of this file is this:
```
{
  "crons": [
    {
      "path": "/api/v1/email/reports/hosting",
      "schedule": "0 */5 * * *"
    }
  ]
}
```
I have also verified this by removing this file from the subsequent commit which triggers the deployment flow as expected.

**Reply by amyegan** [reply]
If you try to [deploy with Vercel CLI](https://vercel.com/docs/cli/deploying-from-cli#) from local, do you see error info? You can use [`vercel --debug`](https://vercel.com/docs/cli/global-options#debug) to make the console output more verbose.

**Reply by yesezra** [reply, marked as answer]
Following up from my post below...I just attempted to deploy with the CLI and did see an error in the debug response:
```
[client-debug] 2023-10-13T21:14:39.581Z Deployment response: {"error":{"code":"cron_jobs_limits_reached","message":"Hobby accounts are limited to daily cron jobs. This cron expression (0 * * * *) would run more than once per day. Upgrade to pro to unlock all Cron Jobs features on Vercel."}}
```

I wasn't even aware that deploying by CLI was possible (I've just been using the GitHub integration since I started using Vercel two weeks ago!) so this is great to know if I encounter issues in the future. It would be great if this error was exposed somewhere with the GitHub integration - if it is exposed somewhere I wasn't able to find it. Thanks @amyegan!

**Comment by yesezra** [comment]
Hey folks,

Like the poster above, I'm attempting to add a cron job to my vercel project. And my project also fails to deploy as soon as I add the vercel.json file. 

I attempted to build a minimal reproducible example for your review, but I wasn't able to replicate the issue – the test case I built as described below does deploy successfully. However, my app still does not deploy when I add a vercel.json file. So I expect that there is something in my app's environment or settings that are causing this issue. Any tips as to how to track that down? 

- Used the [sveltekit boilerplate](https://vercel.com/new/ezra-spiers-projects/templates/svelte/sveltekit-boilerplate) to create a new project. The project was successfully deployed to [sveltekit-cron.vercel.app](https://sveltekit-cron.vercel.app/) and the repo is on GitHub at [https://github.com/yesezra/sveltekit-cron](https://github.com/yesezra/sveltekit-cron)
- added a serverless function at /api/cron using the example code on [this page](https://vercel.com/docs/functions/serverless-functions/quickstart) ([commit](https://github.com/yesezra/sveltekit-cron/commit/f91a8ec14e04480f7040430cc98f31739ef0c9a9))
```javascript
export default function handler(request, response) {
  response.status(200).json({
    body: request.body,
    query: request.query,
    cookies: request.cookies,
  });
}
```
- Verified locally and in production that /api/cron returns the expected result:
<img width="780" alt="Screenshot 2023-10-12 at 8 33 15 PM" src="https://github.com/vercel/community/assets/173440/88fb1ab5-4c37-438e-8c88-28ae19a6ed35">
- Added a `vercel.json` file using the example from [the cron quickstart](https://vercel.com/docs/cron-jobs/quickstart) ([commit](https://github.com/yesezra/sveltekit-cron/commit/90c68e654e38649d5b34295e6584b593aae2a798))

```javascript
{
    "crons": [
      {
        "path": "/api/cron",
        "schedule": "0 5 * * *"
      }
    ]
  }
```
- Verify successful deployment and existence of cron job in the project's settings page.

**Reply by EldinKajdic** [reply]
Hi @yesezra, I just had the same issue and I believe it was due to the cron limitations of the hobby version in Vercel. Not sure what you have, but if you see [this link](https://vercel.com/docs/cron-jobs/usage-and-pricing), you will notice that the hobby version only allows one cron trigger per day. Changing my vercel from this:

```
{
  "crons": [
    {
      "path": "/api/email",
      "schedule": "*/10 * * * *"
    }
  ]
}
```


To this:

```
{
  "crons": [
    {
      "path": "/api/email",
      "schedule": "0 1 * * *"
    }
  ]
}
```


Allowed me to deploy again. Could also be related to the cron expression itself, not sure.

**Reply by aprilmintacpineda** [reply]
That is awful. Like a mad gf, Vercel will completely ignore you and refuse to tell you what's wrong.

**Reply by amyegan** [reply]
Deploying with Vercel CLI is the best way to get error details related to `vercel.json` config problems (see https://github.com/vercel/community/discussions/4424#discussioncomment-7275737). This is because there is no [build log](https://vercel.com/docs/deployments/logs) to see when a config problem prevents the build from starting on [Git integration](https://vercel.com/docs/deployments/git) deployments.

If I run some of these examples now, I see `Hobby accounts are limited to daily cron jobs. This cron expression (0 * * * *) would run more than once per day.`

<img width="674" alt="" src="https://github.com/vercel/community/assets/8247105/010809a8-5db6-40ef-b6fe-4718a663e85c">


Hobby plans currently have a limit of 2 cron jobs which can be triggered 1 time per day: [vercel.com/docs/cron-jobs/usage-and-pricing#usage-&-pricing-for-cron-jobs](https://vercel.com/docs/cron-jobs/usage-and-pricing#usage-&-pricing-for-cron-jobs)

**Reply by aprilmintacpineda** [reply]
That's great, because I actually just moved to using `vercel.json`, I originally commit changes with `export const maxDuration = 900` and I actually got the error, I tested it with `vercel.json` and to my surprise it ignored my commits, would be good as well if vercel does a check on commit so we could see the error right away, otherwise, is there any CLI command I can run to check this manually without triggering a deployment? Like `vercel validate`?

**Comment by yesezra** [comment]
Hey @EldinKajdic - thanks for the tip! I just updated my schedule and it deployed with no issues at all. Hurray! Thanks so much!

To the Vercel folks - it would be great if there was an error somewhere to point to this rather than a silent failure. 

Many thanks!
